### ISSUE TITLE
Add a SOM grid visualization via minisom (reduction + codebook clustering)

### ISSUE THREAD

**Issue by jswest** [jswest/fluster#13]
## Motivation

Right now items are visualized as a UMAP point cloud. I'd like a **grid** visualization too, via a Self-Organizing Map (SOM). [minisom](https://github.com/JustGlowing/minisom) (Python) is the easiest fit; [midsommer](https://github.com/jswest/midsommer) (TS, mine) was the other candidate.

## Decision

After discussion, the design is:

- **SOM is a new reduction method, computed by `minisom` in the Python pipeline** — not in the client. This keeps the SOM stored with receipts, seed-reproducible, and append-only, consistent with the rest of fluster. midsommer stays a visual reference only (and it's a Svelte 4 package vs. our Svelte 5 client).
- **SOM is a reduction with codebook clustering on top** (the classic two-level SOM). The SOM produces grid coordinates *and* a codebook; clustering runs on the codebook and propagates to items via their best-matching unit. This reuses fluster's existing reduction/cluster decoupling rather than inventing a new concept.
- **The viz gets a generalized layout selector** so any 2D reduction (UMAP scatter or SOM grid) can be displayed for a run, independent of which reduction the clustering used.

## Background: how this fits the current architecture

The pipeline is a chain of independent, versioned stages: `materialize → embed → reduce → cluster → exemplars → label → critique`. Two existing properties make this clean:

1. **Reductions and clusterings are already decoupled.** `reduce_items` writes `reductions` + `reduction_coordinates`; `cluster_items` references a reduction *by name* (`ClusteringConfig.reduction = "umap_8d"`) and writes `cluster_runs` + `cluster_assignments`.
2. **The "which coordinates do I draw" choice is already a seam.** `getScatterPlotData` reads `run.reductionId` — today that's the *clustering* reduction (`umap_8d`), so the scatter is really the first two of eight UMAP dims and the dedicated `umap_2d` reduction is effectively unused. The layout selector below leans on (and tidies up) this seam.

A SOM is naturally *both* a reduction (item → grid cell) and a clustering (node → micro-cluster); modeling it as a reduction-with-codebook-clustering captures both without bolting a new stage onto the pipeline.

### Pipeline / data

- **`SOMReduction` in `plan.py`**: grid `x`/`y`, `sigma`, `learning_rate`, `num_iteration`, `topology`, seed (42). Trains on the post-PCA `working` matrix (same input UMAP receives). Must appear after PCA in `plan.reductions`.
- **Item coordinates**: store each item's BMU cell `(i, j)` in `reduction_coordinates` (an honest grid — items stack per cell). `method='som'` requires relaxing the `reductions` CHECK constraint (`method IN ('pca','umap')` → add `'som'`); small migration.
- **Codebook as a first-class receipt**: new table

  ```sql
  som_nodes(reduction_id, node_index, grid_i, grid_j, weight_json, umatrix_dist)
  ```

  storing the node weight vectors (in PCA-50 space) and the U-matrix distance per node. Preferred over fattening `params_json` — queryable and honest.
- **`cluster_items` gains a second mode**: `ClusteringConfig` gets `target: "coordinates" | "codebook"` (default `coordinates`, so nothing existing changes). For `som_2d` + `codebook`: load `som_nodes`, cluster the weight vectors (agglomerative — the classic two-level choice), then each item inherits its BMU node's cluster. Output is ordinary `cluster_assignments`, so exemplars / label / critique downstream are untouched.

### Viz

- **Generalize** `getScatterPlotData` to accept a layout `reductionId` (defaulting to current behavior), plus a `getLayouts()` query returning the project's 2D reductions.
- **Run page** gets a layout dropdown (UMAP scatter / SOM grid); the reduction's `method` selects the component. This also fixes the `umap_8d`-vs-`umap_2d` oddity noted above.
- **New native grid component**: `x×y` cells with U-matrix shading, items colored by the run's cluster, count/stacking per cell, click → `ItemDrawer`. Matches the existing canvas aesthetic (midsommer as reference only).

### Smaller calls

- **Opt-in, not default.** SOM is not added to the default plan (it lengthens every run); it's available config in `plan.yaml` with good defaults when present.
- **Grid-size default** ≈ `5·√N` total nodes, roughly square, configurable.
- **New dependency**: `minisom` in `pyproject.toml`.

## Proposed phases

Each phase gated by reviewer → gitter → check-in.

1. **Plan schema + minisom reduction** — `SOMReduction`, `som_nodes` table + CHECK migration, `reduce_items` produces `som_2d` coords + codebook + U-matrix.
2. **Codebook clustering** — `ClusteringConfig.target`, `cluster_items` codebook mode with BMU propagation.
3. **Layout-selector query refactor** — generalize `getScatterPlotData`, add `getLayouts()`.
4. **Grid component + run page** — native SOM grid view, layout dropdown, README/demo notes.
